llvm/lib/CodeGen/AsmPrinter/CodeViewDebug.cpp

```diff
@@ -2605,6 +2605,9 @@ CodeViewDebug::lowerRecordFieldList(const DICompositeType *Ty) {
       MemberBaseType = TypeTable.writeLeafType(BFR);
     }
     uint64_t MemberOffsetInBytes = MemberOffsetInBits / 8;
+    if (Ty->getName().find("<lambda") != std::string::npos) {
+      MemberName = "__this";
+    }
```

Comment:
This should probably be done by checking whether the `Ty` is a lambda, rather than searching through its name. Oh, this is down in LLVM, so we wouldn't have any info about this.

So maybe this should be up in clang? Hmm, even compiling for a windows target, I don't see a member for the `this` pointer in the LLVM IR metadata - we only get members for the captures, by the looks of it... 

Maybe this MemberInfo stuff is PDB specific/maybe it synthesizes a this pointer from somewhere else... 

I'll leave this to folks more familiar with the PDB side of things.

Reply:
Yeah, I debugged and saw that when parsing the AST the pdb information is generated. Add member information is also here

looking at the lambda function, its naming rules are as follows
https://github.com/llvm/llvm-project/blob/edd690b02e16e991393bf7f67631196942369aed/clang/lib/AST/MicrosoftMangle.cpp#L250-L255

Therefore, if want to modify the place where the information is generated, I don’t know if there is any other way to know that it is a lambda

Reply:
This is inside a for loop iterating the class members. Doesn't this changes all class members name belonging to a lambda to "__this"?

Reply:
Just realized that this fix should be done in Clang:
https://github.com/llvm/llvm-project/blob/0e42df4031e8b2fec357e07ca5ca3b81adf0b5ad/clang/lib/CodeGen/CGDebugInfo.cpp#L1660-L1662

When emitting codeview, use `__this` for the captured `this`.
```
      StringRef ThisName = CGM.getCodeGenOpts().EmitCodeView ? "__this": "this";
      llvm::DIType *fieldType = createFieldType(
          ThisName, type, f->getLocation(), f->getAccess(),
          layout.getFieldOffset(fieldno), VUnit, RecordTy, CXXDecl);
```

Then, the test can be a C++ file put inside `clang/test/CodeGenCXX`.

Reply:
Good, I tried patching it here and it also works, and the pdb compiled by msvc also has operators:
```
 0x465D | LF_FIELDLIST [size = 160, hash = 0x26758]
          - LF_ONEMETHOD [name = `operator()`]
            type = 0x4653, vftable offset = -1, attrs = public
          - LF_METHOD [name = `<lambda_1>`, # overloads = 2, overload list = 0x4658]
          - LF_MEMBER [name = `__this`, Type = 0x464E, offset = 0, attrs = private]
          - LF_MEMBER [name = `aa`, Type = 0x0074 (int), offset = 8, attrs = private]
          - LF_MEMBER [name = `bb`, Type = 0x0074 (int), offset = 12, attrs = private]
          - LF_MEMBER [name = `cc`, Type = 0x0074 (int), offset = 16, attrs = private]
          - LF_ONEMETHOD [name = `operator=`]
            type = 0x465C, vftable offset = -1, attrs = public
          - LF_ONEMETHOD [name = `__autoclassinit2`]
```
Not sure what it means